src/ray/common/scheduling/scheduling_alg.cc

```diff
@@ -0,0 +1,204 @@
+  add(local_node_id, node_resources);
+}
+
+void ClusterResources::add(int64_t node_id, const NodeResources &node_resources) {
+
+  auto it = nodes_.find(node_id);
+  if (it == nodes_.end()) {
+    // This node is new. Add it to the list.
+    nodes_.emplace(node_id, node_resources);
+  } else {
+    // This node exists. Update it's resources.
+    NodeResources &nr = it->second;
+
+    // First, update its predefined resources.
+    for (int i = 0; i < NUM_PREDIFINED_RESOURCES; i++) {
+      nr.capacities[i].total = node_resources.capacities[i].total;
+      nr.capacities[i].available = node_resources.capacities[i].available;
+    }
+
+    // Then, delete existing custom resources and replace them with the new ones.
+    nr.custom_resources.clear();
+
+    auto mcr = node_resources.custom_resources;
+    for (auto it = mcr.begin(); it != mcr.end(); ++it) {
+      nr.custom_resources.insert(*it);
+    }
+  }
+}
+
+bool ClusterResources::remove(int64_t node_id) {
+  auto it = nodes_.find(node_id);
+  if (it == nodes_.end()) {
+    /// Node not found.
+    return false;
+  } else {
+    /// Remobe node.
+    it->second.custom_resources.clear();
+    nodes_.erase(it);
+    return true;
+  }
+}
+
+int64_t ClusterResources::isSchedulable(const TaskReq &task_req, const NodeResources &nr) {
+  int violations = 0;
+
+  /// First, check predefined resources.
+  for (int i = 0; i < NUM_PREDIFINED_RESOURCES; i++) {
+    if (task_req.predefined_resources[i].demand > nr.capacities[i].available) {
+      if (task_req.predefined_resources[i].soft) {
+        // A soft constraint has been violated.
+        violations++;
+      } else {
+        // A hard constraint has been violated.
+        return -1;
+      }
+    }
+  }
+
+  for (int i = 0; i < task_req.custom_resources.size(); i++) {
+    auto it = nr.custom_resources.find(task_req.custom_resources[i].id);
+
+    if (it == nr.custom_resources.end()) {
+      /// Requested resource doesn't exist at this node.
+      if (task_req.custom_resources[i].req.soft) {
+        violations++;
+      } else {
+        return -1;
+      }
+    } else {
+      if (task_req.custom_resources[i].req.demand > it->second.available) {
+        /// Resource constraint is violated.
+        if (task_req.custom_resources[i].req.soft) {
+          violations++;
+        } else {
+          return -1;
+        }
+      }
+    }
+  }
+  return violations;
+}
+
+int64_t ClusterResources::getSchedulableNode(const TaskReq &task_req, int64_t *violations) {
+  /// Min number of violations across all nodes that can schedule the request.
+  int64_t min_violations = INT_MAX;
+  /// Node associated to min_violations.
+  int64_t best_node = -1;
+  *violations = 0;
+
+  /// Check whether local node is schedulable.
+  auto it = nodes_.find(local_node_id_);
+  if (it != nodes_.end()) {
+    int64_t v = isSchedulable(task_req, it->second);
+    if (v == 0) {
+      auto it_p = task_req.placement_hints.find(local_node_id_);
```

Comment:
This logic is not obvious, could use a comment. It seems that we'll only return the local node immediately if it is in the task's placement hints. Shouldn't we also be doing this in the case where there are no placement hints? Or are we defaulting to putting the node in the placement hints somewhere?

Reply:
Yes, we return the local node immediately only if there are no violations. If the local node is not in the placement hints this means a violation.